Add support for ignore_logging_functions in Dynamo to skip arbitrary logging callables during tracing

manigkrish opened the pull request:
Fixes #168317

Adds support for ignoring arbitrary Python callables during TorchDynamo tracing.
Introduces the new `ignore_logging_functions` config and updates DebuggingVariable
so functions in this set are treated as full no-ops without reordering, execution,
or graph breaks. The existing behavior for other logging configurations remains unchanged.

Tested locally by adding functions to `ignore_logging_functions` and verifying
that Dynamo skips them without graph breaks and produces the correct compiled output.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

manigkrish commented:
@pytorchbot label "topic: not user facing"

williamwen42 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -1716,7 +1716,10 @@ def call_function(self, tx: "InstructionTranslator", args, kwargs):
+        # If this function is explicitly configured to be ignored, treat it
+        # as a no-op and do not run it or record it for reordering.
+        if self.value in torch._dynamo.config.ignore_logging_functions:
+            return
```
We should be returning something here - probably `ConstantVariable.create(None)` (and elsewhere in this function)

williamwen42 reviewed (changes requested):
Please also add tests (add to wherever reorderable_logging_functions is tested)

manigkrish commented on torch/_dynamo/config.py:
```diff
@@ -586,6 +586,11 @@ def default_debug_dir_root() -> str:
+# NEW: A set of arbitrary Python callables that should be completely ignored.
+# These functions will NOT be reordered, will NOT run, and will NOT cause
+# graph breaks. They become true no-ops during compilation.
+ignore_logging_functions: set[Callable[..., Any]] = set()
```
Thanks for the suggestion! For this PR, I kept the two configs separate to avoid breaking existing ignore_logger_methods behavior, but I’m open to following a unification direction if that is preferred. Please advise and I can update accordingly.

manigkrish commented:
Hi @williamwen42 

I want to double-check the correct direction before updating the PR, because your two earlier comments seemed to suggest different paths.

Originally you mentioned:
“We’ll need to define a new variable implemented similarly to DebuggingVariable.”

But later you said:
“ignore_logger_methods is only used in LoggingLoggerVariable – I think it’s fine to add additional functionality to ignore_logger_methods.”

To confirm:

Should I still create a new variable (like IgnoredFunctionVariable),
or should I not create a new variable at all and instead implement the behavior entirely inside LoggingLoggerVariable using ignore_logger_methods?

manigkrish committed:
Expand ignore_logger_methods to arbitrary functions via ignore_logging_functions + add IgnoredFunctionVariable + tests

manigkrish commented:
> Let's create a new variable (e.g. `IgnoredFunctionVariable`) defined similarly to `DebugVariable` (i.e. `call_function` doesn't actually trace into the code), but we can re-use `ignore_logger_methods` (renamed to `ignore_logging_functions`) for both `LoggingLoggerVariable` and `IgnoredFunctionVariable`.
> 
> So 1 new variable, but no new config.

Thanks @williamwen42  I’ve applied all requested changes:

• Added IgnoredFunctionVariable
• Reused ignore_logger_methods (renamed to ignore_logging_functions)
• Updated LoggingLoggerVariable
• Updated builder logic
• Added tests where reorderable_logging_functions is tested

All tests pass locally.  
Please let me know if you need any additional changes!

williamwen42 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -1754,9 +1754,26 @@ def can_reorder_logs(fn, args, kwargs) -> True:
+    def call_function(self, tx, args, kwargs):
+        # Correct import
+        from torch._dynamo.variables import ConstantVariable
+        return ConstantVariable.create(None)
```
```suggestion
        return variables.ConstantVariable.create(None)
```

williamwen42 reviewed (commented):
We should also add this config to our docs: i.e. https://docs.pytorch.org/docs/main/compile/programming_model.common_graph_breaks.html#printing-and-logging

In addition to adding `ignore_logging_functions` as a suggestion, we should also mention that `reorderable_logging_functions` has restrictions on function arguments (only tensors, constants, format strings are allowed) and these functions should return None.

manigkrish committed:
dynamo: expand ignore_logger_methods to arbitrary callables via ignore_logging_functions

williamwen42 commented on docs/source/compile/programming_model.common_graph_breaks.md:
```diff
@@ -128,6 +128,10 @@ This config is used to reorder logging functions so that they are called at the
 
+Note: `reorderable_logging_functions` has restrictions, these functions must return `None`, and their arguments must be limited to tensors, constants, or format strings.
+
+You can also use `torch._dynamo.config.ignore_logging_functions` to completely skip selected logging functions during tracing.
```
Can we also include this information about the new flag?
```
# Ignored functions can take any arguments, but MUST return None.
# Functions should either be module-level functions, `logging.Logger.<method>` (ignores all method for all logging.Logger instances) or `logger_obj.<method>` (ignores method only for logger_obj logging.Logger instance).
# Other functions may or may not be ignored due to implementation details. If you want to ignore a function
# that `ignore_logging_functions` is failing to ignore, please submit an issue.
```

manigkrish replied:
Hi @williamwen42 
I’ve applied the requested changes, updated the config alias, improved the documentation, and fixed the lint/formatting issues in the related files.

Thank you for the detailed guidance!

manigkrish committed:
Fix lint and formatting, address review comments

manigkrish committed:
Fix lint, sync registry, address CI failures

manigkrish committed:
Fix whitespace lint failures

manigkrish committed:
CI Failure Fix

Microve commented on docs/source/user_guide/torch_compiler/compile/programming_model.common_graph_breaks.md:
```diff
@@ -128,6 +128,14 @@ This config is used to reorder logging functions so that they are called at the
+
+You can also use `torch._dynamo.config.ignore_logging_functions` to completely skip selected logging functions during tracing.
+
+Ignored functions can take any arguments, but MUST return `None`.
```
> but MUST return `None`

Is this true? but in `test_ignore_function_returns_none`, function `ignore_me` returns "should_not_run"
